src/video/wayland/SDL_waylandwindow.c

```diff
@@ -448,8 +489,105 @@ static const struct xdg_toplevel_listener toplevel_listener_xdg = {
+decoration_frame_configure(struct libdecor_frame *frame,
+                           struct libdecor_configuration *configuration,
+                           void *user_data)
+{
+    SDL_WindowData *wind = user_data;
+    SDL_Window *window = wind->sdlwindow;
+    int width, height;
+    enum libdecor_window_state window_state;
+    struct libdecor_state *state;
+
+    /* window size */
+    if (!libdecor_configuration_get_content_size(configuration, frame, &width, &height)) {
+        width = window->w;
+        height = window->h;
+    }
+
+    wind->resize.width = width;
+    wind->resize.height = height;
+
+    wind->resize.pending = SDL_TRUE;
+    wind->resize.configure = SDL_TRUE;
+    Wayland_HandlePendingResize(window);
+    wind->shell_surface.libdecor.initial_configure_seen = SDL_TRUE;
+
+    window->w = wind->resize.width;
+    window->h = wind->resize.height;
+
+    /* window state */
+    if (!libdecor_configuration_get_window_state(configuration, &window_state)) {
+        window_state = LIBDECOR_WINDOW_STATE_NONE;
+    }
+
+    if (window_state & LIBDECOR_WINDOW_STATE_MAXIMIZED) {
+        window->flags |= SDL_WINDOW_MAXIMIZED;
+    }
+    else {
+        window->flags &= ~SDL_WINDOW_MAXIMIZED;
+    }
+
+    if (window_state & LIBDECOR_WINDOW_STATE_ACTIVE) {
+        window->flags |= SDL_WINDOW_INPUT_FOCUS;
+    }
+    else {
+        window->flags &= ~SDL_WINDOW_INPUT_FOCUS;
+    }
+
+    /* The fullscreen flag is already set my some other entity when 'SDL_SetWindowFullscreen'
+     * is called, but we will set it here again in case the compositor requests fullscreen.
+     */
+    if (window_state & LIBDECOR_WINDOW_STATE_FULLSCREEN) {
+        window->flags |= SDL_WINDOW_FULLSCREEN;
+    }
+    else {
+        window->flags &= ~SDL_WINDOW_FULLSCREEN;
+    }
+
+    /* commit frame state */
+    state = libdecor_state_new(width, height);
+    libdecor_frame_commit(frame, state, configuration);
+    libdecor_state_free(state);
+
+    /* Update the resize capability. Since this will change the capabilities and
+     * commit a new frame state with the last known content dimension, this has
+     * to be called after the new state has been commited and the new content
+     * dimensions were updated. */
+    Wayland_SetWindowResizable(SDL_GetVideoDevice(), window,
+                               window->flags & SDL_WINDOW_RESIZABLE);
+}
+
+static void
+decoration_frame_close(struct libdecor_frame *frame, void *user_data)
+{
+    SDL_SendWindowEvent(((SDL_WindowData *)user_data)->sdlwindow, SDL_WINDOWEVENT_CLOSE, 0, 0);
+}
+
+static void
+decoration_frame_commit(struct libdecor_frame *frame, void *user_data)
+{
+    SDL_WindowData *wind = user_data;
 
+    if (wind->sdlwindow->flags & SDL_WINDOW_OPENGL) {
+        SDL_GL_SwapWindow(wind->sdlwindow);
+    }
+    else if (wind->sdlwindow->flags & SDL_WINDOW_VULKAN) {
+        wl_surface_commit(wind->surface);
+    }
+    else {
+        wl_surface_commit(wind->surface);
+    }
+}
```

Comment:
Should this be like the eglSwapBuffers workaround we have for xdg-shell? Wasn't sure if this was the libdecor equivalent of that function or if this is something different entirely. Either way the SwapWindow call is a bit scary since it sounds like this could be executed mid-frame, which would look pretty bad (if not just outright crash depending on when it happens).

Reply:
libdecor invokes `decoration_frame_commit` when it needs the toplevel surface to commit its state so that the compositor updates the toplevel surface and its subsurfaces (i.e. the decoration subsurfaces). This is required, when the visual of the decorations change, e.g. when the mouse-over changes the button colour or when the client sets a new title.

If the client commits the toplevel anyway (e.g. an `eglSwapBuffers` will do that), then this would not be required. But in general, libdecor needs a way to tell the client that the decoration visuals changed and the client has to commit its toplevel.

Doing it this way is the naive version as it may commit a partial toplevel state. Better would be to synchronise this with the main loop. Ideas?